src/utils.ml

```diff
@@ -105,29 +47,133 @@ module Intf_or_impl = struct
+          Selected_ast.To_ocaml.copy_structure st
+          |> Ocaml_common.Ast_mapper.add_ppx_context_str ~tool_name:"ppx_driver"
+          |> Selected_ast.Of_ocaml.copy_structure
+        in
+        Impl st
+end
+
+module Ast_io = struct
+  type input_version = (module OCaml_version)
+
+  type t = {
+    input_name : string;
+    input_version : input_version;
+    ast : Intf_or_impl.t;
+  }
+
+  type read_error =
+    | Not_a_binary_ast of string
+    | Unknown_version of string
+
+  let magic_length = String.length Ocaml_common.Config.ast_impl_magic_number
+
+  let read_magic ic =
+    let buf = Bytes.create magic_length in
+    let len = input ic buf 0 magic_length in
+    let s = Bytes.sub_string buf ~pos:0 ~len in
+    if len = magic_length then
+      Ok s
+    else
+      Error s
 
-  let to_ast_io (ast : t) ~add_ppx_context =
+  let read_binary_ast ic =
+    match read_magic ic with
+    | Error s -> Error (Not_a_binary_ast s)
+    | Ok s -> (
+        match Find_version.from_magic s with
+        | Intf (module Input_version : OCaml_version) ->
+            let input_name : string = input_value ic in
+            let ast = input_value ic in
+            let module Input_ast = Select_ast (Input_version) in
+            let ast = Intf_or_impl.Intf (Input_ast.Of_ocaml.copy_signature ast) in
+            Ok
+              {
+                input_name;
+                input_version = (module Input_version : OCaml_version);
+                ast;
+              }
+        | Impl (module Input_version : OCaml_version) ->
+            let input_name : string = input_value ic in
+            let ast = input_value ic in
+            let module Input_ast = Select_ast (Input_version) in
+            let ast = Intf_or_impl.Impl (Input_ast.Of_ocaml.copy_structure ast) in
+            Ok
+              {
+                input_name;
+                input_version = (module Input_version : OCaml_version);
+                ast;
+              }
+        | None ->
+            if
+              String.equal s
+                (String.sub Ocaml_common.Config.ast_impl_magic_number ~pos:0
+                  ~len:9)
+              || String.equal s
+                  (String.sub Ocaml_common.Config.ast_intf_magic_number ~pos:0
+                      ~len:9)
+            then Error (Unknown_version s)
+            else Error (Not_a_binary_ast s) )
+
+  let write_as_binary_ast oc
+      { input_name; input_version = (module Input_version); ast } ~add_ppx_context
+      =
+    let module Input_ast = Select_ast (Input_version) in
+    let ast = if add_ppx_context then Intf_or_impl.add_ppx_context ast else ast in
     match ast with
     | Intf sg ->
-      let sg = Selected_ast.To_ocaml.copy_signature sg in
-      let sg =
-        if add_ppx_context then
-          Ocaml_common.Ast_mapper.add_ppx_context_sig ~tool_name:"ppx_driver" sg
-        else
-          sg
-      in
-      Ast_io.Intf sg
+        let sg = Input_ast.To_ocaml.copy_signature sg in
```

Comment:
If `add_ppx_context` is `true`, it seems that:

- we convert from the ppxlib AST version to the OCaml one
- we add the cookies
- then we convert back to the ppxlib version
- then here we convert again to the OCaml one

That seems inefficient. We should skip the last two steps.

Reply:
As we've discussed, I've squashed the last two conversions now if `add_ppx_context` is `true`.